volumina/utility/preferences.py

```diff
@@ -77,37 +77,50 @@ def setmany(self, *args: Tuple[str, str, Any]) -> None:
 
     def read(self):
         try:
-            with self._lock, open(self.path) as f:
-                return json.load(f)
+            with self._lock:
+                return json.loads(self.path.read_text())
         except FileNotFoundError:
             return {}
+        except Exception:
+            logger.exception("Failed to read preferences from %s", self.path)
+            return {}
 
-    def write(self, data):
-        with self._lock:
-            os.makedirs(self.path.parent, exist_ok=True)
-            with open(self.path, "w") as f:
-                json.dump(data, f, indent=2, sort_keys=True)
+    def write(self, data) -> bool:
+        try:
+            with self._lock:
+                os.makedirs(self.path.parent, exist_ok=True)
+                # Touch the file only after successful serialization
+                # in order to prevent existing file corruption.
+                self.path.write_text(json.dumps(data, indent=2, sort_keys=True))
+            return True
+        except Exception:
+            logger.exception("Failed to write preferences to %s", self.path)
+            self.path.unlink()
+            return False
 
-    def migrate(self, old_path):
+    def migrate(self, old_path: Path) -> None:
         if not old_path.exists():
             return
 
         import pickle
+        from volumina.utility import ShortcutManager as SM
 
         try:
-            with open(old_path, "rb") as f:
-                old_preferences = pickle.load(f)
-        except Exception:
-            return
-
-        try:
-            _ = json.dumps(old_preferences)
+            old_preferences = pickle.loads(old_path.read_bytes())
+            # Old pickled preferences saved shortcuts as a dict
+            # with tuple keys, but JSON can only have string keys,
+            # so transform the old dict into the new format.
+            reversemap = old_preferences.get(SM.PreferencesGroup, {}).get(SM.PreferencesSetting, {})
+            if reversemap:
+                old_preferences.setdefault(SM.PreferencesGroup, {})[SM.PreferencesSetting] = [
+                    {"group": group, "name": name, "keyseq": keyseq} for (group, name), keyseq in reversemap.items()
+                ]
         except Exception:
-            logger.warning("Encountered an exception on preferences import - skipping import.")
+            logger.exception("Failed to read old preferences")
             return
 
-        self.write(old_preferences)
-        old_path.unlink()
+        if self.write(old_preferences):
+            old_path.unlink()
```

Comment:
Also here, renaming to `ilastik_preferences.old` or something would be nice, in case users want to go back to the previous versions (I am constantly jumping through our versions, and with  this I loose all my settings for the old version).

Reply:
We can just leave the old file as-is here too. New versions read from a new locations anyway.

This way, new ilastik versions will look at the old location, transform/copy those preferences to the new location at startup, and use only the new location, while old versions will just use the old location as if nothing happened.

Reply:
I thought about this more and looking at the code, this does not seem to be the case. `migrate` is always executed when starting ilastik. This should _only_ be done if there is no _new_ preferences file. If we do it like that, the two can coexist in isolation, which I think is a good enough solution.

Reply:
I've changed the migration _exit_ condition to `self.path.exists() and not old_path.exists()`, that is, we actually migrate only if we didn't do it before (there is no new file) and we have something to migrate from (the old file exists).

Reply:
that's perfect! if you remove the test assertion that old file is gone, this can be merged!